x/airdrop/keeper/claim.go

```diff
@@ -0,0 +1,195 @@
+	}
+
+	// Confirm the user has not elected to claim early (meaning their type will be ClaimDaily)
+	if userAllocation.ClaimType != types.CLAIM_DAILY {
+		return types.ErrClaimTypeUnavailable.Wrapf("user has already elected claim option %s",
+			userAllocation.ClaimType.String())
+	}
+
+	// Confirm we're not past the decision date
+	currentTime := ctx.BlockTime().Unix()
+	if currentTime > airdrop.ClaimTypeDeadlineDate.Unix() {
+		return types.ErrAfterDecisionDeadline
+	}
+
+	// Confirm the airdrop started
+	if currentTime < airdrop.DistributionStartDate.Unix() {
+		return types.ErrDistributionNotStarted
+	}
+
+	// Sum the total rewards 0 them out in the process
+	totalAccruedRewards := sdkmath.ZeroInt()
+	for i, rewardsOnDate := range userAllocation.Allocations {
+		totalAccruedRewards = totalAccruedRewards.Add(rewardsOnDate)
+		userAllocation.Allocations[i] = sdkmath.ZeroInt()
+	}
+
+	// If there are no rewards, alert the user with an error
+	if totalAccruedRewards.IsZero() {
+		return types.ErrNoUnclaimedRewards
+	}
+
+	// Calculate rewards after claim early penalty
+	rewardsRemainingRate := sdk.OneDec().Sub(airdrop.EarlyClaimPenalty)
+	distributedRewards := sdk.NewDecFromInt(totalAccruedRewards).Mul(rewardsRemainingRate).TruncateInt()
+
+	// Update the amount claimed on the allocation record
+	// claimed += distributedRewards
+	userAllocation.Claimed = userAllocation.Claimed.Add(distributedRewards)
+
+	// Update the amount forfeited on the allocation record
+	// forfeited += totalAccruedRewards - distributedRewards
+	// Note: forfeited should be zero before the next operation,
+	// but we're doing += just in case there's a scenario where it's not zero in the future
+	forfeitedRewards := totalAccruedRewards.Sub(distributedRewards)
+	userAllocation.Forfeited = userAllocation.Forfeited.Add(forfeitedRewards)
+
+	// Flag the user's claim type decision
+	userAllocation.ClaimType = types.CLAIM_EARLY
+
+	// Distribute rewards from the distributor, deducting the early penalty
+	distributorAccount := sdk.MustAccAddressFromBech32(airdrop.DistributionAddress)
+	claimerAccount := sdk.MustAccAddressFromBech32(userAllocation.Address)
+
+	rewardsCoin := sdk.NewCoin(airdrop.RewardDenom, distributedRewards)
+	if err := k.bankKeeper.SendCoins(ctx, distributorAccount, claimerAccount, sdk.NewCoins(rewardsCoin)); err != nil {
+		return errorsmod.Wrapf(err, "unable to distribute rewards")
+	}
+
+	// Update the reward record for to mark the progress
+	k.SetUserAllocation(ctx, userAllocation)
+
+	return nil
+}
+
+// Admin transaction to merge allocations between a stride and non-stride address
+// If the stride address does not yet have an allocation, the host allocation will be overwritten
+// with the stride address
+// Otherwise, if the stride allocation already exists, the two allocations will be merged and set
+// in on the stride allocation
+// We can safely change the type back to DAILY because if a user claimed early, their allocations
```

Comment:
hmm that's a very good question

Reply:
I think I'd prefer to go with the simplest approach, since this situation shouldn't actually happen in practice (because our frontend requires users to link before they claim). So I think I'd lean towards leaving it as is since I think that would be the simpler implementation

Reply:
That makes sense, and agreed with the simpler implementation (particularly as this is an edge case). Resolving!